### PR TITLE
Enforce Java 8 in Gradle builds

### LINKED ISSUE #14499: gradle build should enforce minimal jvm version
Its very unintuitive today: this happens often on accident because i switch JAVA_HOME to java 7 when backporting (and sometimes forget to switch it back). 

But in general, users may not be aware java 8 is required. Currently exceptions look like this:
```
:buildSrc:build UP-TO-DATE
=======================================
Elasticsearch Build Hamster says Hello!
=======================================
  Gradle Version : 2.8
  JDK Version    : 1.7.0_55-b13 (Oracle Corporation)
  OS Info        : Linux 4.2.0-16-generic (amd64)
:core:compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':core:compileJava'.
> invalid flag: -Xdoclint:all/private

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED
```

### PR DESCRIPTION
**Pull request by jasontedor**
This commit forces builds with Gradle to require at least Java 8.

Closes #14499

### CODE REVIEW

**Comment by jasontedor** [comment]
This gives:

```
11:39:21 [jason:~/src/elastic/elasticsearch] enforce-java-8 ± JAVA_HOME=`/usr/libexec/java_home -v 1.7` gradle assemble | pbcopy

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/jason/src/elastic/elasticsearch/core/build.gradle' line: 24

* What went wrong:
A problem occurred evaluating project ':core'.
> Failed to apply plugin [id 'elasticsearch.build']
   > Java 8 or above is required to build Elasticsearch

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
```

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy [hunk 1]
```diff
@@ -59,6 +59,11 @@ class BuildPlugin implements Plugin<Project> {
                 throw new GradleException('Gradle 2.6 or above is required to build elasticsearch')
             }
 
+            // enforce Java version
+            if (JavaVersion.current().compareTo(JavaVersion.VERSION_1_8) < 0) {
```
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: Groovy makes comparison operators available to enums. So you can just do:

```
if (JavaVersion.current() < JavaVersion.VERSION_1_8) {
```

**Comment by rjernst** [comment]
LGTM, one tiny nit.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy (hunk 1 above)
Comment (rmuir) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: or just use the isCompatibleWith. No need to use esoteric groovy language features, it makes plenty of bad choices.

Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: +1 to use isCompatibleWith

Comment (jasontedor) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: @rmuir @rjernst As in b0c70650007221e7465ba8207d71100e9e9637a3?

Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: +1

**Pull request merged**